tests/suite.py

```diff
@@ -0,0 +1,132 @@
+            interpolation=configparser.ExtendedInterpolation(),
+            delimiters=(':'),
+            comment_prefixes=('#')
+        )
+        config.read(cls.config_file)
+
+        cls.suite_root = os.path.split(os.path.abspath(cls.config_file))[0]
+        cls.timeout = config.getfloat("General", "timeout", fallback=1.0)
+
+        if 'variables' in config and 'paths' in config:
+            intersecting_keys = set(config["paths"].keys())\
+                                .intersection(set(config["variables"].keys()))
+            if len(intersecting_keys) > 0:
+                raise ValueError(
+                    "The sections 'paths' and 'variables' must not share keys, "
+                    "but they have the following common key{:s}: {:s}"
+                    .format(
+                        's' if len(intersecting_keys) > 1 else '',
+                        ', '.join(k for k in intersecting_keys)
+                    )
+                )
+
+        if 'variables' in config:
+            for key in config['variables']:
+                setattr(cls, key, config['variables'][key])
+
+        if 'paths' in config:
+            for key in config['paths']:
+                # assert(key not in cls._globals)
+                rel_path = config['paths'][key]
+                abs_path = os.path.abspath(
+                    os.path.join(cls.suite_root, rel_path)
+                )
+                if not os.path.exists(abs_path):
+                    raise ValueError(
+                        "Path replacement for {short}: {abspath} does not exist"
+                        " (was expanded from {rel})".format(
+                            short=key,
+                            abspath=abs_path,
+                            rel=rel_path)
+                    )
+                setattr(cls, key, abs_path)
+
+    def test_run(self):
+
+        dct = _disjoint_dict_merge(self.__dict__, type(self).__dict__)
+        cwd = os.path.split(inspect.getfile(type(self)))[0]
+
+        for command, retval, stdout, stderr in zip(self.commands,
+                                                   self.retval,
+                                                   self.stdout,
+                                                   self.stderr):
+            command, retval, stdout, stderr = map(
+                lambda s: str(s).format(**dct),
+                [command, retval, stdout, stderr]
+            )
+            retval = int(retval)
+            success = True
+            timeout = False
+
+            proc = subprocess.Popen(
+                shlex.split(command),
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                cwd=cwd
+            )
+
+            def timeout_reached():
+                global success, timeout
+                success = False
+                timeout = True
+                proc.kill()
+
+            t = threading.Timer(self.timeout, timeout_reached)
```

Comment:
This is purely to have a portable way how to stop the running process after the timeout has expired. The problem is that `subprocess.Popen` did not have the `timeout` parameter itself until very recently (especially not in Python2). But if you use the threading module, you can stop a running process via `process.kill()` from a different thread. That's where the threading module comes in, as the Timer runs in a different thread.

While it would be nice to run the tests in parallel, I am afraid Python's unittest framework does not support that.